endgame_simulations/endgame_simulation.py

```diff
@@ -0,0 +1,216 @@
+            >>> [sample.mf_prevalence_in_population() for sample in simulation.iter_run(end_time=3, sampling_interval=1.0)]
+            [0.99, 0.6, 0.2]
+
+        Args:
+            end_time (float): end time
+            sampling_interval (float): State sampling interval (years)
+
+        Yields:
+            Iterator[State]: Iterator of the simulation's state.
+        """
+        ...
+
+    @overload
+    def iter_run(
+        self, *, end_time: float, sampling_years: list[float]
+    ) -> Iterator[State]:
+        """Run the simulation until `end_time`. Generates stats for every year in `sampling_years`.
+
+        This is a generator, so you must it as one.
+
+        Examples:
+            >>> simulation = Simulation(start_time=0, params=Params(), n_people=400)
+            >>> for state in simulation.iter_run(end_time=10, sampling_years=[0.1, 1, 5])
+            ...    print(state.mf_prevalence_in_population())
+            0.99
+            0.6
+            0.2
+
+        Args:
+            end_time (float): end time
+            sampling_years (list[float]): list of years to sample State
+
+        Yields:
+            Iterator[State]: Iterator of the simulation's state.
+        """
+        ...
+
+    def iter_run(
+        self,
+        *,
+        end_time: float,
+        sampling_interval: float | None = None,
+        sampling_years: list[float] | None = None,
+    ) -> Iterator[State]:
+        if end_time < self.state.current_time:
+            raise ValueError("End time after start")
+
+        if sampling_interval and sampling_years:
+            raise ValueError(
+                "You must provide sampling_interval, sampling_years or neither"
+            )
+
+        if sampling_years:
+            sampling_years = sorted(sampling_years)
+
+        sampling_years_idx = 0
+
+        with tqdm.tqdm(
+            total=end_time - self.state.current_time + self._delta_time,
+            disable=not self.verbose,
+        ) as progress_bar:
+            while self.state.current_time <= end_time:
+                is_on_sampling_interval = (
+                    sampling_interval is not None
+                    and self.state.current_time % sampling_interval < self._delta_time
+                )
+
+                is_on_sampling_year = (
+                    sampling_years
+                    and sampling_years_idx < len(sampling_years)
+                    and abs(
+                        self.state.current_time - sampling_years[sampling_years_idx]
+                    )
+                    < self._delta_time
+                )
+
+                if is_on_sampling_interval or is_on_sampling_year:
+                    yield self.state
+                    if is_on_sampling_year:
+                        sampling_years_idx += 1
+
+                progress_bar.update(self._delta_time)
+                type(self).advance_state(self.state, self.debug)
```

Comment:
I think we are making an assumption here that advance_state will change current time by self._delta_time (otherwise, progress bar doesn't reflect the reality). We might as well add the time advancement here, instead of having it in advance_state.

Reply:
Fair point - perhaps we can assume timestep for now